File: Simulation/15683번 - 감시/15683.py
import sys
import copy

input = sys.stdin.readline

n, m = map(int, input().split())
arr = []
cctv = []
for i in range(n):
    arr_idx = list(map(int, input().split()))
    for j in range(m):
        if arr_idx[j] != 0 and arr_idx[j] != 6:
            cctv.append([i, j])
    arr.append(arr_idx)
ans = 10**7

def view(x, y, d):
    d %= 4
    v = [[0, -1], [-1, 0], [0, 1], [1, 0]]
    dx = v[d][0]
    dy = v[d][1]
    
    while(1):
        x += dx
        y += dy
        if 0 <= x < n and 0 <= y < m and arr[x][y] != 6:
            if arr[x][y] != 0:
                continue
            arr[x][y] = -1
        else:
            break

def dfs(cnt):
    global arr, ans
    
    if cnt == len(cctv):
        zero = 0
        for i in range(n):
            for j in range(m):
                if arr[i][j] == 0:
                    zero += 1
        ans = min(ans, zero)
        return
        
    tmp = copy.deepcopy(arr)
    x = cctv[cnt][0]
    y = cctv[cnt][1]
    for dd in range(4):
        if arr[x][y] == 1:
            view(x, y, dd)
        elif arr[x][y] == 2:
            view(x, y, dd)
            view(x, y, dd + 2)
        elif arr[x][y] == 3:
            view(x, y, dd)
            view(x, y, dd + 1)
        elif arr[x][y] == 4:
            view(x, y, dd)
            view(x, y, dd + 1)
            view(x, y, dd + 2)
        elif arr[x][y] == 5:
            view(x, y, dd)
            view(x, y, dd + 1)
            view(x, y, dd + 2)
            view(x, y, dd + 3)
            
        dfs(cnt + 1)
        arr = copy.deepcopy(tmp)

def solve():
    dfs(0)
    print(ans)
    
if __name__ == '__main__':
    solve()

File: Simulation/15683번 - 감시/15683.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int n, m;
int map[9][9];
vector<pair<int, int> > cctv;
int dx[4] = {0, -1, 0, 1};
int dy[4] = {-1, 0, 1, 0};
int ans = 1e9;

void input() {
    cin >> n >> m;
    
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            cin >> map[i][j];
            if(map[i][j] != 0 && map[i][j] != 6) cctv.push_back({i, j});
        }
    }
}

void see(int x, int y, int d) {
    d %= 4;
    
    while(1) {
        int nx = x + dx[d];
        int ny = y + dy[d];
        x = nx;
        y = ny;
        
        if(nx < 0 || nx >= n || ny < 0 || ny >= m || map[nx][ny] == 6) break;
        if(map[nx][ny] != 0) continue;
        map[nx][ny] = -1;
    }
}

void dfs(int cnt) {
    if(cnt == cctv.size()) {
        int zero = 0;
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(map[i][j] == 0) zero++;
            }
        }
        
        ans = min(ans, zero);
        
        return;
    }
    
    int tmp[9][9];
    int x = cctv[cnt].first;
    int y = cctv[cnt].second;
    
    for(int dir = 0; dir < 4; dir++) {
        copy_n(&map[0][0], 9 * 9, &tmp[0][0]);
        
        if(map[x][y] == 1) {
            see(x, y, dir);
        }
        else if(map[x][y] == 2) {
            see(x, y, dir);
            see(x, y, dir + 2);
        }
        else if(map[x][y] == 3) {
            see(x, y, dir);
            see(x, y, dir + 1);
        }
        else if(map[x][y] == 4) {
            see(x, y, dir);
            see(x, y, dir + 1);
            see(x, y, dir + 2);
        }
        else if(map[x][y] == 5) {
            see(x, y, dir);
            see(x, y, dir + 1);
            see(x, y, dir + 2);
            see(x, y, dir + 3);
        }
        
        dfs(cnt + 1);
        copy_n(&tmp[0][0], 9 * 9, &map[0][0]);
    }
}

void solve() {
    dfs(0);
    
    cout << ans << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;    
}